feat(chips): Add keyboard navigation

patrickrodee opened the pull request:
Add support for keyboard navigation for chips. Update ripple focus-within mixin behavior.

Demos: [action chips](http://chips-keyboard-navigation-demo.surge.sh/spec/mdc-chips/classes/action.html), [choice chips](http://chips-keyboard-navigation-demo.surge.sh/spec/mdc-chips/classes/choice.html), [filter chips](http://chips-keyboard-navigation-demo.surge.sh/spec/mdc-chips/classes/filter.html), and [input chips](http://chips-keyboard-navigation-demo.surge.sh/spec/mdc-chips/classes/input.html).

Fixes #2259

BREAKING CHANGE: Chips markup, adapters, foundations, and events have changed.

moog16 commented on packages/mdc-chips/chip/adapter.ts:
```diff
@@ -107,7 +114,42 @@ export interface MDCChipAdapter {
+  hasTrailingIcon(): boolean;
+
+  /**
+   * Returns true if the primary action content has focus.
```
This says returns true - actually I think this is an incorrect description. Please also add `@param`

moog16 commented on packages/mdc-chips/chip/foundation.ts:
```diff
@@ -190,12 +227,104 @@ export class MDCChipFoundation extends MDCFoundation<MDCChipAdapter> {
+    const chipIsRTL = this.adapter_.isRTL();
+    const chipHasTrailingIcon = this.adapter_.hasTrailingIcon();
+    const focusSource = this.getFocusSource_();
+    const keyIsLeft = key === strings.ARROW_LEFT_KEY && !chipIsRTL || key === strings.ARROW_RIGHT_KEY && chipIsRTL;
```
instead of `keyIsleft` or `keyIsRight`, can you change it to increment/decrement?

patrickrodee commented on packages/mdc-chips/README.md:
```diff
@@ -296,6 +323,7 @@ Event Name | `event.detail` | Description
 `MDCChip:selection` | `{chipId: string, selected: boolean}` | Indicates the chip's selection state has changed (for choice/filter chips)
 `MDCChip:removal` | `{chipId: string, root: Element}` | Indicates the chip is ready to be removed from the DOM
 `MDCChip:trailingIconInteraction` | `{chipId: string}` | Indicates the chip's trailing icon was interacted with (via click/tap or Enter key)
+`MDCChip:navigation` | `{chipId: string, key: string, source: FocusSource}` | Indicates a navigation event has occurred on a chip
```
There are other keydown events that we capture that don't trigger focus navigation. Any arrow key or home/end keydown is considered a navigation event.

moog16 replied:
Can we define what a navigation event is in the Readme

patrickrodee commented on packages/mdc-chips/chip-set/adapter.ts:
```diff
@@ -43,4 +45,24 @@ export interface MDCChipSetAdapter {
   setSelected(chipId: string, selected: boolean): void;
+
+  /**
+   * @return the numerical index of the chip with the matching id or -1.
```
Done

patrickrodee commented on packages/mdc-chips/chip-set/adapter.ts:
```diff
@@ -43,4 +45,24 @@ export interface MDCChipSetAdapter {
+  getIndexOfChipById(chipId: string): number;
+
+  /**
+   * Gives focus to the chip at the given index with the given key.
```
Done

moog16 commented on packages/mdc-chips/chip-set/foundation.ts:
```diff
@@ -100,8 +106,50 @@ export class MDCChipSetFoundation extends MDCFoundation<MDCChipSetAdapter> {
    * Handles the event when a chip is removed.
    */
   handleChipRemoval(chipId: string) {
+    const index = this.adapter_.getIndexOfChipById(chipId);
```
Yesterday we found the issue where removing a chip made 2 chips have tabIndex=0. This can be fixed by setting all tabIndex=-1, and then setting the next chip to tabIndex=0. I think this is the method that would contain that logic.

patrickrodee replied:
Great find, thanks! I've fixed this.

moog16 commented on packages/mdc-chips/README.md:
```diff
@@ -40,10 +40,17 @@ npm install @material/chips
-  <button class="mdc-chip">
-    <span class="mdc-chip__text">Chip content</span>
-  </button>
+<div class="mdc-chip-set" role="grid">
```
Can we also add a note somewhere that in IE11 and Edge the keyboard navigation appears to not work, since mdc-state's focuswithin pseudoselector doesn't work.

patrickrodee replied:
Sure thing

patrickrodee replied:
Done

moog16 commented on docs/framework-wrappers.md:
```diff
@@ -21,6 +21,5 @@ Material Components for the web are architected to be adaptable to various major
     - [Material Components Vue](https://github.com/matsp/material-components-vue): MDC Web Integration for Vue.js (using [vanilla components](./integrating-into-frameworks.md#the-simple-approach-wrapping-mdc-web-vanilla-components))
     - [BalmUI](https://material.balmjs.com/): Next Generation Material UI for Vue.js
     - [Ember Material Components](https://github.com/onehilltech/ember-cli-mdc): MDC Web integration for Ember (using [vanilla components](./integrating-into-frameworks.md#the-simple-approach-wrapping-mdc-web-vanilla-components))
-    - [MatBlazor](https://github.com/SamProf/MatBlazor): Material Design components for Blazor
```
never even heard of blazor lol

patrickrodee commented on packages/mdc-chips/chip/constants.ts:
```diff
@@ -21,23 +21,52 @@
  */
 
+export enum EventSource {
+  Primary,
```
There seem to be no lint rules about it. Let me know if you'd like me to update it to be uppercase.

patrickrodee replied:
Updated!

patrickrodee commented on packages/mdc-chips/chip-set/component.ts:
```diff
@@ -102,14 +108,32 @@ export class MDCChipSet extends MDCComponent<MDCChipSetFoundation> {
     // To ensure we don't accidentally omit any methods, we need a separate, strongly typed adapter variable.
     const adapter: MDCChipSetAdapter = {
+      focusChipAtIndex: (index, key, source) => {
+        if (index < 0 || index > this.chips_.length - 1) {
```
Done!

patrickrodee commented on packages/mdc-chips/chip/foundation.ts:
```diff
@@ -34,6 +34,22 @@ const emptyClientRect = {
   width: 0,
 };
 
+export const NAVIGATION_KEYS = new Set<string>();
```
Done!

patrickrodee commented:
> Found a minor bug in the screenshot demos where pressing Right Arrow when on the last chip would remove the focus from that chip which makes all chip to lose focus. But expected at least one chip to have focus.

Ah nice, thanks. I've fixed that and added some tests for it.

abhiomkar commented on packages/mdc-chips/chip/foundation.ts:
```diff
@@ -189,15 +206,142 @@ export class MDCChipFoundation extends MDCFoundation<MDCChipAdapter> {
+    this.adapter_.setTrailingActionAttr(strings.TAB_INDEX, '-1');
+  }
+
+  focusAction(key: string, source: EventSource) {
```
Since chip set already has key and source information, why do we not delegate this logic to chipset? Chip would just need to know where to focus (Primary or Trailing icon).

patrickrodee replied:
That's a good question. We still need this logic somewhere, whether it's in the chip or the chip set. I'm flexible on where it goes. Let me know what you think.

abhiomkar replied:
Params for this method seems confusing. It is accepting event information similar to `handle*` methods but it is a proxy to component API. I feel we need to move this logic to chip set and use Chip's low level APIs to focus the right element.

Something like:
- `chip.focusPrimary()`
- `chip.focusTrailingIcon()`


WDYT?

patrickrodee replied:
That was my thinking too. Sounds good. I'll update with that.

patrickrodee replied:
Done

patrickrodee commented on packages/mdc-chips/README.md:
```diff
@@ -194,24 +213,32 @@ chipSet.listen('MDCChip:removal', function(event) {
-<button class="mdc-chip mdc-chip--selected">
-  <span class="mdc-chip__text">Add to calendar</span>
-</button>
+<div class="mdc-chip-set mdc-chip-set--choice">
```
Done, thanks!
